### PR TITLE
feat: fetch schemas and generate line diff

### PR DESCRIPTION
**Pull request by Shridhad**
relates: https://github.com/neondatabase/cloud/issues/11846

```
neonctl branches schema-diff [base-branch] <compare-branch> --database <db-name> --projectId <project-id>
```

- `base-branch` is optional, If not specified the command will use the branch set with `set-context` or the primary branch from the project.
- `compare-branch` is compare branch Id or name
- `--database` is optional. If not specified, the command will display diff for all databases for `base-branch`

## Tasks

- [x] Fetch list of all databases for base branches (for role name)
- [x] Fetch schemas for both base and compare branches
- [x] Generate the unified diff and colorize    

https://github.com/neondatabase/neonctl/assets/1949852/0cd457a7-636c-422f-82ff-b2f10e3a9d9c

### CODE REVIEW

**Comment by duskpoet** [comment]
I like the output format, but I think the command itself is too verbose. What about: `neonctl branches diff branch1 branch2`?

File: src/commands/schema_diff.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  ]);
+
+  const patch = createPatch(
+    props.database,
+    baseSchema,
+    compareSchema,
+    `Base Branch: ${baseBranch}`,
+    `Compare Branch: ${compareBranch}`,
+  );
+  writer(props).text(colorize(patch));
```
Comment (duskpoet) [inline src/commands/schema_diff.ts:73]: is there an option to support json/yaml output? Maybe library can return something like an array of diffs?

File: src/commands/branches.ts [hunk 2]
```diff
@@ -186,7 +187,14 @@ export const builder = (argv: yargs.Argv) =>
       'Get a branch',
       (yargs) => yargs,
       async (args) => await get(args as any),
-    );
+    )
+    .command({
+      command: 'schema-diff [branch] <compare-branch>',
```
Comment (duskpoet) [inline src/commands/branches.ts:192]: I think square brackets are of optional args, and optional shouldn't precede required

File: src/commands/schema_diff.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
+import { BranchScopeProps } from '../types';
+import { createPatch } from 'diff';
+import { Database } from '@neondatabase/api-client';
+import chalk from 'chalk';
+import { writer } from '../writer.js';
+import { branchIdFromProps, branchIdResolve } from '../utils/enrichers.js';
+
+type SchemaDiffProps = BranchScopeProps & {
+  branch?: string;
+  baseBranch?: string;
```
Comment (duskpoet) [inline src/commands/schema_diff.ts:10]: you can name the parameter to "branch" to avoid confusions

Comment (duskpoet) [reply src/commands/schema_diff.ts:10]: hmm, i see that is is named "branch", what is "baseBranch" then

File: src/commands/schema_diff.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  compareBranch: string;
+  database: string;
```
Comment (duskpoet) [inline src/commands/schema_diff.ts:12]: I don't see database in command options

File: src/commands/branches.ts (hunk 2 above)
Comment (duskpoet) [reply src/commands/branches.ts:192]: yea, ok, I admit your position is stronger :) 👍

File: src/commands/schema_diff.ts (hunk 3 above)
Comment (duskpoet) [reply src/commands/schema_diff.ts:10]: Why not call it "base-branch" in options?

File: src/commands/schema_diff.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  added: chalk.green,
+  removed: chalk.red,
+  header: chalk.yellow,
+  section: chalk.magenta,
+};
+
+type ColorId = keyof typeof COLORS;
+
+export const schemaDiff = async (props: SchemaDiffProps) => {
+  props.branch = props.baseBranch || props.branch;
```
Comment (duskpoet) [inline src/commands/schema_diff.ts:25]: do you still need this line?

Comment (duskpoet) [reply src/commands/schema_diff.ts:25]: ah, because of context?

File: src/commands/schema_diff.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  baseDatabases.map(async (database) => {
+    const patch = await createSchemaDiff(
+      baseBranch,
+      compareBranch,
+      database,
+      props,
+    );
+    writer(props).text(colorize(patch));
```
Comment (duskpoet) [inline src/commands/schema_diff.ts:67]: not sure if this is important, but the order of output here is random

File: src/commands/schema_diff.ts (hunk 3 above)
Comment (Shridhad) [reply src/commands/schema_diff.ts:10]: Corrected it in command.

**Review by duskpoet — APPROVED** [review]

File: src/commands/schema_diff.ts (hunk 5 above)
Comment (Shridhad) [reply src/commands/schema_diff.ts:25]: yes, for the context. Maybe I can move this line to middleware in next PR

File: src/commands/schema_diff.ts (hunk 6 above)
Comment (Shridhad) [reply src/commands/schema_diff.ts:67]: yes, the order is not important

**Pull request merged**